### PR TITLE
Combine AD and non-AD classes

### PR DESCRIPTION
**Pull request by jasondhales**
Combine AD and non-AD versions of Density and Pressure.

See #15915

### CODE REVIEW

File: framework/src/interfaces/Coupleable.C
Comment (dschwen) [inline framework/src/interfaces/Coupleable.C:2161]: Is.. Is that .. a Lambda?!

File: modules/tensor_mechanics/src/bcs/Pressure.C
Comment (dschwen) [inline modules/tensor_mechanics/src/bcs/Pressure.C:276]: De we even need to implement those specializations? We'd get a runtime linker error if anything tried to use them.

Comment (jasondhales) [reply modules/tensor_mechanics/src/bcs/Pressure.C:276]: That's what I thought.  Maybe I did something wrong, but the tests all died until I added those.

File: modules/tensor_mechanics/src/bcs/Pressure.C
Comment (tophmatthews) [inline modules/tensor_mechanics/src/bcs/Pressure.C:30]: Glad to see you're not just blasting the AD use of `constant`, but should one be deprecated since it's duplicative?

File: modules/tensor_mechanics/include/bcs/Pressure.h
Comment (GiudGiud) [inline modules/tensor_mechanics/include/bcs/Pressure.h:42]: these should still override, they are not member function templates

Comment (GiudGiud) [reply modules/tensor_mechanics/include/bcs/Pressure.h:42]: or is it not parsing it with the std::conditional

File: modules/tensor_mechanics/src/bcs/Pressure.C
Comment (tophmatthews) [reply modules/tensor_mechanics/src/bcs/Pressure.C:30]: `constant` makes more sense in my head since it goes along the convention of function, pp, etc, but don't care strongly

File: modules/tensor_mechanics/src/bcs/Pressure.C
Comment (jasondhales) [reply modules/tensor_mechanics/src/bcs/Pressure.C:276]: Even if I comment out the final two lines (`template class ...`), I get a 'symbol not found' error unless I include those implementations.

File: modules/tensor_mechanics/include/bcs/Pressure.h
Comment (GiudGiud) [reply modules/tensor_mechanics/include/bcs/Pressure.h:42]: afaik it s on a per routine basis

Comment (jasondhales) [reply modules/tensor_mechanics/include/bcs/Pressure.h:42]: The compiler will give a warning if you've used override once and could have used it again.  That's what I was trying to say.  But that doesn't work here because of different base classes.

File: modules/tensor_mechanics/src/bcs/Pressure.C
Comment (jasondhales) [reply modules/tensor_mechanics/src/bcs/Pressure.C:279]: Any thoughts?  I don't see a way around this.

Comment (dschwen) [reply modules/tensor_mechanics/src/bcs/Pressure.C:279]: Yeah, I wanted to check something today. And then I'll either approve/merge or make a suggestion.

**Comment by dschwen** [comment]
Ok, I have a solution for this... hang on...

**Comment by dschwen** [comment]
Take a look at 82510ebcb3 @jasondhales 

`Pressure` is now its own class that inherits from `PressureTempl<false>` and adds the non-AD only methods. (rather than making `Pressure` just a typedef pointing to `PressureTempl<false>`)

Benefits:
- We can use `override`
- We don't have to provide dummy specializations that only contain a `mooseError`

**Review by dschwen — DISMISSED** [review]
:+1: 

Unfortunately there is a small merge conflict now. I'll re-approve once that's fixed.

**Comment by dschwen** [comment]
Also, @loganharbour , this PR now contains a commit by me. So we should have an additional reviewer. There is a design for combined AD/non-AD residual objects (which have different sets of member functions) that allows the use of `override`. This idea is applicable broadly. You should review, too.

**Comment by loganharbour** [comment]
Fix the conflict and then I'll approve?

**Comment by jasondhales** [comment]
@dschwen, I pulled in your commit and sorted out the merge conflict.

**Review by dschwen — APPROVED** [review]

**Pull request merged**
